torchqdynamics/odeint.py

```diff
@@ -1,80 +1,124 @@
-    elif sensitivity == 'autograd':
-        return odeint_main(qsolver, y0, tsave)
-    elif sensitivity == 'adjoint':
-        return odeint_adjoint(qsolver, y0, tsave, parameters)
 
+class GradientAlgorithm(Enum):
+    AUTOGRAD = 'autograd'  # gradient computed by torch
+    NONE = 'none'  # don't compute the gradients
+    ADJOINT = 'adjoint'  # compute the gradient using the adjoint method
+
+
+def odeint(
+    qsolver,
+    y0,
+    t_save: torch.Tensor,
+    exp_ops: List[torch.Tensor] = None,
+    save_states: bool = True,
+    gradient_algorithm=GradientAlgorithm.AUTOGRAD,
+):
+    # check arguments
+    check_t(t_save)
 
-def odeint_main(qsolver, y0, tsave):
+    # dispatch to appropriate odeint subroutine
+    params = (qsolver, y0, t_save, exp_ops, save_states)
+    if gradient_algorithm == GradientAlgorithm.NONE:
+        return _odeint_inplace(*params)
+    elif gradient_algorithm == GradientAlgorithm.AUTOGRAD:
+        return _odeint_main(*params)
+    elif gradient_algorithm == GradientAlgorithm.ADJOINT:
+        return _odeint_adjoint(*params)
+    else:
+        raise ValueError(f'Gradient algorithm {gradient_algorithm} not defined')
+
+
+def _odeint_main(qsolver, y0, t_save, exp_ops, save_states):
     if isinstance(qsolver.options, FixedStep):
-        return _fixed_odeint(qsolver, y0, tsave)
+        return _fixed_odeint(
+            qsolver, y0, t_save, qsolver.options.dt, exp_ops, save_states
+        )
     elif isinstance(qsolver.options, AdaptativeStep):
-        return _adaptive_odeint(qsolver, y0, tsave)
+        return _adaptive_odeint(qsolver, y0, t_save, exp_ops, save_states)
 
 
-def odeint_inplace(qsolver, y0, tsave):
+# For now we use *args and **kwargs for helper methods that are not implemented to ease the potential API
+# changes that could occur later. When a method is implemented the methods should take the same arguments
+# as all others
+def _odeint_inplace(*args, **kwargs):
     # TODO: Simple solution for now so torch does not store gradients. This
     #       is probably slower than a genuine in-place solver.
     with torch.no_grad():
-        return odeint_main(qsolver, y0, tsave)
+        return _odeint_main(*args, **kwargs)
 
 
-def odeint_adjoint(qsolver, y0, tsave, parameters):
+def _odeint_adjoint(*_args, **_kwargs):
     raise NotImplementedError
 
 
-def _adaptive_odeint(qsolver, y0, tsave):
+def _adaptive_odeint(*_args, **_kwargs):
     raise NotImplementedError
 
 
-def _fixed_odeint(qsolver, y0, tsave):
-    # initialize save tensor
-    ysave = torch.zeros(len(tsave), *y0.shape).to(y0)
+def _fixed_odeint(qsolver, y0, t_save, dt, exp_ops, save_states):
+    # Initialize save tensor
+    y_save = None
+    if save_states:
+        y_save = torch.zeros(len(t_save), *y0.shape).to(y0)
+
+    if len(exp_ops) > 0:
+        exp_save = torch.zeros(len(t_save), len(exp_ops)).to(
+            device=y0.get_device(), dtype=torch.float
+        )
+    else:
+        exp_save = None
```

Comment:
It would be better to place `exp_save = None` just above the `if` like you did with `y_save`.